### DISCUSSION TITLE
Thoughts about increasing CDC transfer speed.

### DISCUSSION THREAD

**Discussion by kwagyeman** [hathach/tinyusb#2707] [General]
### Related area

CDC Transfer Speed

### Hardware specification

imxrt

### Is your feature request related to a problem?

Hi, we currently use TinyUSB on the IMXRT for USB CDC.

One of the things I noticed with our USB sniffer with USB HS is that many packets are NACKed.

![image](https://github.com/hathach/tinyusb/assets/5694981/d482ad42-a187-4fab-b4a5-2f5ae9d4a4a4)

In the example above, each 1-2 poll is a NACK. This is wasting USB bandwidth. Ideally, we'd want to have the data ready for USB before it needs it.

![image](https://github.com/hathach/tinyusb/assets/5694981/bb36e440-d9cb-4a8a-b1c8-6e65e3be75a1)

However, there's very little time to setup the data buffer before the USB controller needs the next one. Like 400ns.

Any thoughts on how this could be improved? Right now we are getting a callback via USB-CDC to prepare the data and memcpy it into a buffer. If this callback were executed before the host needed the data then this would improve our bandwidth.

Not sure how this could be done.

### Describe the solution you'd like

Just opening a discussion on how to improve this. We are not bottlenecked by it right now.

https://github.com/openmv/openmv/issues/2261

### I have checked existing issues, dicussion and documentation

- [X] I confirm I have checked existing issues, dicussion and documentation.

**Comment by HiFiPhile** [comment]
>  I confirm I have checked existing issues, discussion and documentation.

Please use search function, there are already many discussion on this topic.

**Reply by iabdalkader** [reply]
Could you point us to one of those discussions related to this topic ? I've searched this repo a lot and I can't find something useful other than your open issue from 2021 https://github.com/hathach/tinyusb/issues/920

**Reply by HiFiPhile** [reply]
If you search with `CFG_TUD_CDC_EP_BUFSIZE` you will get more information.

Generally speaking increase FIFO size `CFG_TUD_CDC_RX_BUFSIZE` and `CFG_TUD_CDC_TX_BUFSIZE` and endpoint buffer size `CFG_TUD_CDC_EP_BUFSIZE` can improve a lot throughput.

However there is is a catch for `CFG_TUD_CDC_EP_BUFSIZE` if host sending data with a length equal to endpoint size. 
Supposing it set to 2048, once a 512b packet is received it will keep waiting instead handle the data to user. A ZLP or short packet is needed to end this transfer.

Currently there is no support for bypassing the FIFO yet.

**Reply by iabdalkader** [reply]
Hi @HiFiPhile I found some related discussions. I will experiment with increasing the buffers and see if it improves the bandwidth. Thank you!

**Reply by iabdalkader** [reply]
> However there is is a catch for CFG_TUD_CDC_EP_BUFSIZE if host sending data with a length equal to endpoint size.
Supposing it set to 2048, once a 512b packet is received it will keep waiting instead handle the data to user. A ZLP or short packet is needed to end this transfer.

This sounds like something OS stacks should handle, right ?

**Reply by HiFiPhile** [reply]
> This sounds like something OS stacks should handle, right ?

I thought it should be, but at least on Windows no zlp is sent by the host.

**Reply by iabdalkader** [reply]
Thanks for the help @HiFiPhile

**Comment by kwagyeman** [comment]
This was solved. Setting CFG_TUD_CDC_EP_BUFSIZE and the RX/TX buffers to the same size using 4096 solved the problem. No more NACKs.

**Reply by kwagyeman** [reply]
Need to verify the fix works on mac/linux/windows before making it official. However, it doesn't increase bandwidth by about 3x.
